File: io/io/src/test/resources/compat/issue-756/src/java/LegacyBinaryStreamCaller.java
package io.bluetape4k.io.serializer.compat.issue756.java;

import io.bluetape4k.io.serializer.BinarySerializer;
import java.io.ByteArrayOutputStream;
import java.io.OutputStream;
import java.lang.reflect.Method;
import java.nio.charset.StandardCharsets;

public final class LegacyBinaryStreamCaller {
    private LegacyBinaryStreamCaller() {
    }

    public static void main(String[] args) throws Exception {
        BinarySerializer serializer = new LegacyBinaryImplementation();
        require("binary".equals(new String(serializer.serialize("binary"), StandardCharsets.UTF_8)),
            "legacy binary call changed");

        Method method = BinarySerializer.class.getMethod(
            "serializeBinaryToStream",
            Object.class,
            OutputStream.class
        );
        require(method.isDefault(), "binary stream method is not a default");

        ByteArrayOutputStream target = new ByteArrayOutputStream();
        Object written = method.invoke(serializer, "binary", target);
        require(Integer.valueOf(6).equals(written), "unexpected binary stream write count");
        require("binary".equals(target.toString(StandardCharsets.UTF_8)), "unexpected binary stream payload");
        System.out.println("legacy-binary-java-stream-caller=PASS");
    }

    private static void require(boolean condition, String message) {
        if (!condition) {
            throw new AssertionError(message);
        }
    }
}

File: io/io/src/test/resources/compat/issue-756/src/kotlin/LegacyBinaryStreamCaller.kt
package io.bluetape4k.io.serializer.compat.issue756.kotlin

import io.bluetape4k.io.serializer.BinarySerializer
import java.io.ByteArrayOutputStream
import java.io.OutputStream

private class LegacyBinaryKotlinImplementation: BinarySerializer {
    override fun serialize(graph: Any?): ByteArray = graph?.toString()?.encodeToByteArray() ?: byteArrayOf()

    @Suppress("UNCHECKED_CAST")
    override fun <T: Any> deserialize(bytes: ByteArray?): T? =
        bytes?.takeIf { it.isNotEmpty() }?.decodeToString() as T?
}

fun main() {
    val serializer: BinarySerializer = LegacyBinaryKotlinImplementation()
    check(serializer.serialize("binary").decodeToString() == "binary") { "legacy binary call changed" }

    val method = BinarySerializer::class.java.getMethod(
        "serializeBinaryToStream",
        Any::class.java,
        OutputStream::class.java,
    )
    check(method.isDefault) { "binary stream method is not a default" }

    val target = ByteArrayOutputStream()
    val written = method.invoke(serializer, "binary", target)
    check(written == 6) { "unexpected binary stream write count" }
    check(target.toString(Charsets.UTF_8) == "binary") { "unexpected binary stream payload" }
    println("legacy-binary-kotlin-stream-caller=PASS")
}
